### PR TITLE
Use atomic transactions when altering databases

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Although it includes unittests it doesn't include an integration test. I would like to test that crashing the node while it's busy in `addValidatedBlock` would ensure the Node's databases don't get corrupted. In theory this should just work, but a test would be nice..

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
I should mention: I had an alternative implementation where each class would have `begin / commit / rollback`, but then I realized there is a lot of code duplication everywhere. It's easier if we just used a managed database class that keeps track of all database handles and then use them all atomically via SQL transactions.

**Comment by TrustHenry** [comment]
I saw it in Forum.dlang. How about this?
Should we always consider `commit` timing?
```
void transaction(Callable)(Callable c) if(isSomeFunction!Callable) {
  db.begin();
  scope(success) db.commit();
  scope(failure) db.rollback();
  c();
 }
```

**Comment by AndrejMitrovic** [comment]
> Should we always consider commit timing?

Commit should only be issued once we're done with all the databases. Additionally if you look closer, I commit as soon as all the databases were modified and before `readLastBlock` is called or `onValidatorsChanged` is called. In case either of these two calls fail I don't want to roll back the UTXO set. So here `scope (success)` would be inappropriate.

File: source/agora/common/ManagedDatabase.d [hunk 1]
```diff
@@ -78,4 +80,81 @@ public class ManagedDatabase
...
+    {
+        foreach (db; thread_dbs)
+        {
+            try
+            {
+                db.begin();
+            }
+            catch (Exception exc)
+            {
+                // in most cases this would only trigger if there was a
```
Comment (AndrejMitrovic) [inline source/agora/common/ManagedDatabase.d:112]: I will fix this up with the work that's going on in https://github.com/bpfkorea/agora/issues/836. I need all the SQLite errors to properly be handled in the binding, and then consider which cases here are fatal /  could be retried.

File: source/agora/common/ManagedDatabase.d [hunk 2]
```diff
@@ -78,4 +80,81 @@ public class ManagedDatabase
...
+        Begins a batch operation on all the open databases.
+
+        A beginBatch() call must be followed by either a commitBatch() or
+        a rollback() call (in case of exceptions).
+
+        The rollback() call will cancel any transactions started after the
+        beginBatch() call.
+
+        In case of crashes it is not necessary to call rollback() because
+        none of the transactions were COMMITED, and since the application
```
Comment (AndrejMitrovic) [inline source/agora/common/ManagedDatabase.d]:
```suggestion
        none of the transactions were commited, and since the application
```

**Comment by Geod24** [comment]
Regarding commit 1: Working on eliminating this function
Regarding commit 2: Doesn't that just switch the problem ? Those two operations shouldn't affect each other.

The rest, needs a bit more review (today).

**Comment by AndrejMitrovic** [comment]
> Regarding commit 2: Doesn't that just switch the problem ? Those two operations shouldn't affect each other.

You explained to me IRL that we should always store a block regardless if adding stuff to the UTXO set fails?

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
Needs a rebase tho

**Pull request merged**